# Computational Genomic Analysis

>
Genome-wide mapping of cancer dependency genes and genetic modifiers of chemotherapy using a high-risk hepatoblastoma-like genetic model
>

## Abstract
>>
A lack of relevant genetic models and cell lines hampers our understanding of hepatoblastoma pathogenesis and the development of 
new therapies for this neoplasm. We report an improved liver-specific MYC-driven hepatoblastoma-like murine model that faithfully 
recapitulates the pathological features of embryonal hepatoblastoma, with transcriptomics resembling the high-risk gene signatures 
of the human disease.  Single-cell RNA-sequencing and spatial transcriptomics identified distinct subpopulations of hepatoblastoma 
cells. After deriving cell lines from the mouse model, we mapped the cancer dependency genes using CRISPR-Cas9 screening and identified 
druggable targets shared with human hepatoblastoma (i.e., CDK7, CDK9, PRMT1, PRMT5). Our screen also revealed oncogenes and tumor 
suppressor genes in hepatoblastoma that engage multiple, druggable cancer signaling pathways. Chemotherapy is critical for human 
hepatoblastoma treatment. A genetic mapping of doxorubicin response by CRISPR-Cas9 screening identified modifiers whose loss-of-function 
synergizes with (e.g., PRKDC) or antagonizes (e.g., apoptosis genes) with the effect of chemotherapy. The combination of PRKDC 
inhibition and doxorubicin-based chemotherapy greatly enhances therapeutic efficacy. These studies have provided a useful set of 
resources including disease models suitable to identify and validate potential therapeutic targets in human high-risk hepatoblastoma.
>>


## Corresponding authors

[Xiang Chen](https://www.stjude.org/directory/c/xiang-chen.html),
[Jun Yang](https://www.stjude.org/research/labs/yang-jun-lab.html)
	
2023, St. Jude Children's Research Hospital
